src/celldega/sp/__init__.py

```diff
@@ -0,0 +1,359 @@
+    polygon_geom = gdf_polygon.geometry.iloc[0]
+
+    gdf_points_with_dist = gdf_points.copy()
+    gdf_points_with_dist[f"distance_to_polygon_{roi_name}"] = gdf_points_with_dist.distance(polygon_geom)
+    print (f"Distances from points to polygon {roi_name} were calculated.")
+    return gdf_points_with_dist, gdf_polygon
+
+def plot_distance_to_roi(
+    gdf_points: gpd.GeoDataFrame,
+    gdf_polygon: gpd.GeoDataFrame,
+    roi_name: str,
+    cmap=cm.plasma,
+    markersize: float = 0.5
+):
+    """
+    Plot the points colored by their distance to the region of interest.
+    
+    Args:
+        gdf_points: GeoDataFrame containing points with distance column.
+        gdf_polygon: GeoDataFrame containing the target polygon.
+        roi_name: Name of the region of interest.
+        cmap: Colormap for distance visualization.
+        markersize: Size of the points in the plot.
+    """
+
+    print (f"Plotting distances from points to polygon {roi_name}...")
+
+    # Normalize distances for color mapping
+    norm = mcolors.Normalize(
+        vmin=gdf_points[f"distance_to_polygon_{roi_name}"].min(),
+        vmax=gdf_points[f"distance_to_polygon_{roi_name}"].max(),
+    )
+    
+    # Map distances to colors
+    gdf_points["color"] = gdf_points[f"distance_to_polygon_{roi_name}"].apply(
+        lambda x: cmap(norm(x))
+    )
+    
+    # Identify points inside the roi
+    polygon = gdf_polygon.geometry.iloc[0]  # Assuming one roi
+    gdf_points["inside_polygon"] = gdf_points["geometry"].apply(
+        lambda x: polygon.contains(x))
+    
+    # Create the plot
+    fig, ax = plt.subplots(figsize=(20, 8))
+    
+    # Plot points outside the roi with gradient color
+    gdf_points[~gdf_points["inside_polygon"]].plot(
+        ax=ax,
+        color=gdf_points[~gdf_points["inside_polygon"]]["color"],
+        aspect=1,
+        legend=True,
+        markersize=markersize,
+    )
+    
+    # Plot the roi as a filled area with label
+    gdf_polygon.plot(
+        ax=ax, color="gray", aspect=1, alpha=0.5, edgecolor="black", linewidth=2
+    )
+    
+    # Annotate the polygon with the ROI name at its centroid
+    centroid = gdf_polygon.geometry.iloc[0].centroid
+    ax.annotate(
+        text=roi_name,
+        xy=(centroid.x, centroid.y),
+        ha="center",
+        va="center",
+        fontsize=12,
+        color="black",
+    )
+    
+    # Add colorbar for the gradient
+    sm = cm.ScalarMappable(cmap=cmap, norm=norm)
+    sm.set_array([])
+    cbar = plt.colorbar(sm, ax=ax, orientation="vertical", pad=0.01, aspect=40)
+    cbar.set_label(f"Distance to ROI {roi_name} (micron)")
+    
+    # Adjust the plot
+    ax.invert_yaxis()
+    plt.title(
+        f"Region Selected Cells Colored Based on Distance to the ROI {roi_name} (micron)"
+    )
+    plt.tight_layout()
+    plt.xlabel("x (micron)")
+    plt.ylabel("y (micron)")
+    plt.show()
+
+
+def create_concentric_rings(polygon, num_bands, band_width):
```

Comment:
It would be nice to not have to say how many bands we want but instead have this done automatically. One way to do this would be to first calculate the alpha shape of the entire tissue and then calculate a large number of rings (way more than we need, ~100), then only keep the rings that overlap with the alpha shape of the tissue (you could even take the intersection of the tissue-alpha-shape and ring if you want).

Reply:
to reduce computing and the depandency on alphashapes, I will just compare with a gdf_reference which can be gdf of tissue cells that are readily available.